File: kotlin/provenance-mark/src/main/kotlin/com/blockchaincommons/provenancemark/Xoshiro256StarStar.kt
package com.blockchaincommons.provenancemark

class Xoshiro256StarStar private constructor(private val s: ULongArray) {
    init {
        require(s.size == 4) { "state must have 4 elements" }
    }

    fun toState(): ULongArray = s.copyOf()

    fun toData(): ByteArray {
        val data = ByteArray(32)
        for (i in 0 until 4) {
            val value = s[i]
            val offset = i * 8
            for (j in 0 until 8) {
                data[offset + j] = ((value shr (j * 8)) and 0xFFu).toByte()
            }
        }
        return data
    }

    fun copy(): Xoshiro256StarStar = fromState(s)

    fun nextULong(): ULong {
        val result = (s[1] * 5u).rotateLeft(7) * 9u
        val t = s[1] shl 17

        s[2] = s[2] xor s[0]
        s[3] = s[3] xor s[1]
        s[1] = s[1] xor s[2]
        s[0] = s[0] xor s[3]

        s[2] = s[2] xor t
        s[3] = s[3].rotateLeft(45)

        return result
    }

    fun nextByte(): Byte = (nextULong() and 0xFFu).toByte()

    fun nextBytes(length: Int): ByteArray = ByteArray(length) { nextByte() }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Xoshiro256StarStar) return false
        return s.contentEquals(other.s)
    }

    override fun hashCode(): Int = s.contentHashCode()

    companion object {
        fun fromState(state: ULongArray): Xoshiro256StarStar {
            require(state.size == 4) { "state must have 4 elements" }
            return Xoshiro256StarStar(state.copyOf())
        }

        fun fromData(data: ByteArray): Xoshiro256StarStar {
            require(data.size == 32) { "data must be 32 bytes" }
            val state = ULongArray(4)
            for (i in 0 until 4) {
                var value = 0uL
                val offset = i * 8
                for (j in 0 until 8) {
                    value = value or ((data[offset + j].toUByte().toULong()) shl (j * 8))
                }
                state[i] = value
            }
            return Xoshiro256StarStar(state)
        }
    }
}

private fun ULong.rotateLeft(bits: Int): ULong =
    (this shl bits) or (this shr (64 - bits))

File: csharp/ProvenanceMark/ProvenanceMark/Xoshiro256StarStar.cs
using System.Buffers.Binary;
using System.Numerics;

namespace BlockchainCommons.ProvenanceMark;

/// <summary>
/// Deterministic xoshiro256** PRNG used by provenance-mark.
/// </summary>
public sealed class Xoshiro256StarStar : IEquatable<Xoshiro256StarStar>
{
    private readonly ulong[] _state;

    private Xoshiro256StarStar(ulong[] state)
    {
        _state = state;
    }

    public ulong[] ToState() => (ulong[])_state.Clone();

    public byte[] ToData()
    {
        var data = new byte[32];
        for (var index = 0; index < 4; index++)
        {
            BinaryPrimitives.WriteUInt64LittleEndian(data.AsSpan(index * 8, 8), _state[index]);
        }
        return data;
    }

    public Xoshiro256StarStar Copy() => FromState(_state);

    public ulong NextUInt64()
    {
        var result = BitOperations.RotateLeft(_state[1] * 5, 7) * 9;
        var t = _state[1] << 17;

        _state[2] ^= _state[0];
        _state[3] ^= _state[1];
        _state[1] ^= _state[2];
        _state[0] ^= _state[3];

        _state[2] ^= t;
        _state[3] = BitOperations.RotateLeft(_state[3], 45);

        return result;
    }

    public byte NextByte() => (byte)(NextUInt64() & 0xff);

    public byte[] NextBytes(int length)
    {
        var bytes = new byte[length];
        for (var index = 0; index < length; index++)
        {
            bytes[index] = NextByte();
        }
        return bytes;
    }

    public static Xoshiro256StarStar FromState(ReadOnlySpan<ulong> state)
    {
        if (state.Length != 4)
        {
            throw new ArgumentException("state must have 4 elements", nameof(state));
        }

        return new Xoshiro256StarStar(state.ToArray());
    }

    public static Xoshiro256StarStar FromData(ReadOnlySpan<byte> data)
    {
        if (data.Length != 32)
        {
            throw new ArgumentException("data must be 32 bytes", nameof(data));
        }

        var state = new ulong[4];
        for (var index = 0; index < 4; index++)
        {
            state[index] = BinaryPrimitives.ReadUInt64LittleEndian(data[(index * 8)..]);
        }

        return new Xoshiro256StarStar(state);
    }

    public bool Equals(Xoshiro256StarStar? other)
    {
        return other is not null && _state.AsSpan().SequenceEqual(other._state);
    }

    public override bool Equals(object? obj) => Equals(obj as Xoshiro256StarStar);

    public override int GetHashCode()
    {
        var hash = new HashCode();
        foreach (var value in _state)
        {
            hash.Add(value);
        }
        return hash.ToHashCode();
    }
}
